### PR TITLE
JDK-8245058: improve presentation of annotations for modules and packages

### PR DESCRIPTION
**Pull request by hns**
This PR adds a dedicated signature to module and package overview pages and moves all code for generating signatures to a new `doclets.formats.html.Signatures` utility class. The `Signatures` class contains two public static methods for module and package signatures as well as two inner classes for type and member signatures. 

Signatures are rendered as `<div>` elements with top level CSS classes that identify the kind of signature: `module-signature`, `package-signature`, `type-signature` and `member-signature`. Signature components are rendered as `<span>` elements with CSS classes that identify the component, e.g.: `modifiers`, `annotations`, `element-name`, `exceptions`. 

As a side benefit, this PR reduces the number of methods in `HtmlDocletWriter` used to generate annotation information.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux additional | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- |
| Build | ⏳ (8/8 running) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ⏳ (9/9 running) | ❌ (1/9 failed) | ⏳ (9/9 running) | ⏳ (8/9 running) |

**Failed test task**
- [Linux x86 (jdk/tier1 part 3)](https://github.com/hns/jdk/runs/1465127426)

### Issue
 * [JDK-8245058](https://bugs.openjdk.java.net/browse/JDK-8245058): improve presentation of annotations for modules and packages

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to bd2ef09b70dcd7085c39681456433cf231e15615

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/966/head:pull/966`
`$ git checkout pull/966`

### CODE REVIEW

**Comment by hns** [comment]
API docs:
http://cr.openjdk.java.net/~hannesw/8245058/api.01/index.html
Example module signature: 
http://cr.openjdk.java.net/~hannesw/8245058/api.01/java.base/module-summary.html
Example package signature:
http://cr.openjdk.java.net/~hannesw/8245058/api.01/java.base/java/lang/package-summary.html

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java [hunk 1]
```diff
@@ -1678,68 +1678,23 @@ public boolean isCoreClass(TypeElement typeElement) {
     }
 
     /**
-     * Adds the annotation types for the given packageElement.
+     * Return a content tree containing the  annotation types for the given Element.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:1681]: (minor) either put type names (Element) in code font or use English equivalent (element).

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java [hunk 2]
```diff
@@ -1678,68 +1678,23 @@ public boolean isCoreClass(TypeElement typeElement) {
...
      *
-     * @param packageElement the package to write annotations for.
-     * @param htmltree the documentation tree to which the annotation info will be
+     * @param element an Element
+     * @param lineBreak if true add new line between each member value
+     * @return the documentation tree to which the annotation info will be
      *        added
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:1686]: The method is `getAnnotationInfo` so it seems strange/confusing to return an object to which the info _will be_ added (future tense.)  I'm surprised it's not a documentation tree containing the annotation info.
Is this a hangover from the old code?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java [hunk 3]
```diff
@@ -1678,68 +1678,23 @@ public boolean isCoreClass(TypeElement typeElement) {
...
-     *        added
-     */
-    private void addAnnotationInfo(List<? extends AnnotationMirror> descList, Content htmltree) {
-        htmltree.add(getAnnotationInfo(descList, true));
+    Content getAnnotationInfo(Element element, boolean lineBreak) {
+        return getAnnotationInfo(element.getAnnotationMirrors(), lineBreak);
     }
 
     /**
      * Return a content tree containing the annotation types for the given element.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:1693]: `Returns`

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java [hunk 4]
```diff
@@ -1760,11 +1715,11 @@ Content getAnnotationInfo(List<? extends AnnotationMirror> descList, boolean lin
      * the given doc.
      *
      * @param descList a list of annotation mirrors.
-     * @param linkBreak if true, add new line between each member value.
+     * @param lineBreak if true, add new line between each member value.
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:1718]: Funny typo ;-)  Good catch!

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlStyle.java [hunk 5]
```diff
@@ -531,51 +531,77 @@
     // The following constants are used for the components of a signature of an element
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlStyle.java:531]: Two lines earlier, line 529, change `member signature` to `signatures`

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/resources/stylesheet.css [hunk 6]
```diff
@@ -485,9 +485,6 @@ div.table-tabs > button.table-tab {
 /*
  * Styles for contents.
  */
-.description pre {
-    margin-top:0;
-}
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/resources/stylesheet.css:490]: Yay! Good improvement.

**Review by jonathan-gibbons — APPROVED** [review]
Generally, very nice to see this continuation of the improvements to handle signatures.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java [hunk 7]
```diff
@@ -194,110 +194,12 @@ protected TypeElement getCurrentPageElement() {
...
-            Content link = getLink(new LinkInfoImpl(configuration, LinkInfoImpl.Kind.RECORD_COMPONENT,
-                    e.asType()));
-            content.add(link);
-            content.add(Entity.NO_BREAK_SPACE);
-            content.add(e.getSimpleName());
-            sep = ", ";
-        }
-        content.add(")");
-        return content;
-    }
```
[the remark this answers is not in the document]
Comment (hns) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/ClassWriterImpl.java:300]: This method is part of the type signature generating code, it has been moved to Signatures.TypeSIgnature without functional changes.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java (hunk 2 above)
Comment (hns) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:1686]: Indeed, the method returns the fully rendered content. I'm fixing this in a follow-up commit (along with other issues).

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlStyle.java (hunk 5 above)
Comment (hns) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/markup/HtmlStyle.java:531]: Good catch...

**Pull request closed without merging**